src/wp-includes/media.php

```diff
@@ -1059,12 +1059,31 @@ function wp_get_attachment_image( $attachment_id, $size = 'thumbnail', $icon = f
 		 */
 		$context = apply_filters( 'wp_get_attachment_image_context', 'wp_get_attachment_image' );
 
-		// Add `loading` attribute.
-		if ( wp_lazy_loading_enabled( 'img', $context ) ) {
-			$default_attr['loading'] = wp_get_loading_attr_default( $context );
+		$attr = wp_parse_args( $attr, $default_attr );
+
+		$loading_optimization_attr = wp_get_loading_optimization_attributes(
+			'img',
+			// $attr doesn't contain width and height by default.
+			array_merge(
+				array(
+					'width'  => $width,
+					'height' => $height,
+				),
+				$attr
+			),
+			$context
+		);
+
+		// Remove `loading` optimization attribute if not enabled.
+		if ( ( isset( $loading_optimization_attr['loading'] ) && ! wp_lazy_loading_enabled( 'img', $context ) ) ||
```

Comment:
@joemcgill I tend to agree with you, however there is a potential (temporary) caveat: Trying to explain my thoughts:
* We could introduce a function like `wp_maybe_strip_loading_attribute( $loading_attrs, $context )` that would just unset `$loading_attrs['loading']` if lazy-loading for the given `$context` is disabled. That function could be called right before returning the result of `wp_get_loading_optimization_attributes()`. Generally, I'd say that's much better than calling the logic everywhere that `wp_get_loading_optimization_attributes()` is called.
* However, the way that `wp_get_loading_optimization_attributes()` is written uses lots of early returns, so we would need to add that call pretty much everywhere in those returns, in several of which we already heavily call `wp_maybe_add_fetchpriority_high_attr()`, so we would end up with lots of `return wp_get_loading_optimization_attributes( wp_maybe_add_fetchpriority_high_attr( $loading_attrs, $attr ), $context );` lines, which is just a bit odd as well.
* I would like to avoid refactoring the function to get rid of early returns since the logic is quite complicated, so if we wanted to do such a refactoring it should happen in a PR just for that purpose, to not make review of this PR here more complicated.
* Alternatively to the above chaining of function calls, we could introduce a filter and hook both functions in it, but that would still mean that `apply_filters()` call would need to be added to every early return statement.

All of that is probably not a big deal, but I also don't really like any of those solutions, but definitely don't want to get into refactoring the function logic to avoid early returns in this PR. Curious to hear your thoughts. Maybe the approach of chaining the functions in `wp_get_loading_optimization_attributes()` is still a better trade-off than handling it outside of `wp_get_loading_optimization_attributes()`? Since then at least we would only have a potential iteration to do _within_ the function which has far fewer risks for BC breaks.

@kt-12 What do you think?

Reply:
Even though I agree with @joemcgill that this code should have been within the `wp_get_loading_optimization_attributes`. I would have preferred refactoring  the code ( i did try it early to see if it is easily possible.,.) but like @felixarntz pointed out the logic is a bit complex. The only benefit in chaining I see is that in future we will only need to change one function apart from that i don't see any other benefit. Personally, I prefer the current state of code as it is less complex.

Reply:
@kt-12 Thinking about it further, I actually think the approach of chaining the 1-2 function calls in every `return` statement in `wp_get_loading_optimization_attributes()` is the best for now, for the following reasons:
* Allows us to refactor later without breaking any backward compatibility, which would not be possible if we kept the function calls outside of `wp_get_loading_optimization_attributes()`.
* Avoids the need for refactoring right now.

It's not the nicest solution when looking at the code, but that trade-off is worth it compared to future BC breaks. If we keep the function calls outside of `wp_get_loading_optimization_attributes()`, we would later have to change the scope of what that function does if we refactor it, which would be risky in terms of BC.